Work towards faster dispatch

pgkeller opened the pull request:
Make launch_msg dispatch copy more robust
Add null kernels env var

pgkeller committed:
#4620: Make launch_msg dispatch copy more robust

pgkeller committed:
#4620: Add env var TT_METAL_NULL_KERNELS

DrJessop commented on tt_metal/hw/firmware/src/brisck.cc:
```diff
@@ -26,11 +26,13 @@ uint8_t noc_index = NOC_INDEX;
     kernel_profiler::mark_time(CC_KERNEL_MAIN_START);
     kernel_main();
     kernel_profiler::mark_time(CC_KERNEL_MAIN_END);
+#endif
```
Can we clean this up to pass in defines only to worker programs for DEBUG_NULL_KERNELS and not send to dispatch core? That way we don't need extra DISPATCH_KERNEL define.

pgkeller replied:
I don't know of a mechanism that informs the build of the difference betwen a worker program and the dispatch programs.  Seems that would be uglier than this from 10K feet.

pgkeller replied:
could rename "DISPATCH_KERNEL" to "INFRA_KERNEL" (or some such) to make the mechanism more generic

DrJessop replied:
Given that you made a directed test, maybe this ifdef can be commited via a separate PR or at the very least it's own commit referencing the new skip kernel debug mode issue.

pgkeller committed:
#4620: Add new test: test_pgm_dispatch_perf

Disables kernel execution and dumps runtime of dispatch cost
Tweakable kernel size, #cores, #cbs, #rt args

DrJessop commented on tests/tt_metal/tt_metal/test_pgm_dispatch_perf.cpp:
```diff
@@ -0,0 +1,166 @@
+    tt::llrt::OptionsG.set_kernels_nullified(false);
+
+    return 0;
+}
```
Can this be in a gtest?

DrJessop commented on tests/tt_metal/tt_metal/test_pgm_dispatch_perf.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        for (int i = 0; i < n_cbs_g; i++) {
+            tt_metal::CircularBufferConfig cb_config = tt_metal::CircularBufferConfig(16, {{i, tt::DataFormat::Float16_b}})
+                .set_page_size(i, 16);
```
This page size corresponds to the size of the page data, not the size of the config. I guess in our case, it can be anything since the kernels don't actually run.

DrJessop commented on tests/tt_metal/tt_metal/test_pgm_dispatch_perf.cpp:
```diff
@@ -0,0 +1,166 @@
+#include "tt_metal/llrt/rtoptions.hpp"
+
+constexpr uint32_t DEFAULT_ITERATIONS = 10;
+constexpr uint32_t MAX_KERNEL_SIZE = 8192;
```
Let's allow for up to 16KB.

DrJessop commented on tests/tt_metal/tt_metal/test_pgm_dispatch_perf.cpp:
```diff
@@ -0,0 +1,166 @@
+        log_info(LogTest, "Ran in {}us", elapsed_seconds.count() * 1000 * 1000);
+        log_info(LogTest, "Ran in {}us per iteration", elapsed_seconds.count() * 1000 * 1000 / iterations_g);
+
+        pass &= tt_metal::CloseDevice(device);
```
I believe gtest has benchmarking.

DrJessop commented on tests/tt_metal/tt_metal/module.mk:
```diff
@@ -37,6 +37,7 @@ TT_METAL_TESTS += \
 		 tests/tt_metal/test_matmul_multi_core_multi_dram_in0_mcast \
 		 tests/tt_metal/test_matmul_multi_core_multi_dram_in1_mcast \
 		 tests/tt_metal/test_matmul_multi_core_multi_dram_in0_mcast_in1_mcast \
+		 tests/tt_metal/test_pgm_dispatch_perf \
```
Let's move your code to a gtest, these are legacy tests.

DrJessop reviewed (changes requested):
Looks great, but let's move your test to be a gtest.
